staging/src/k8s.io/sample-controller/controller.go

```diff
@@ -0,0 +1,301 @@
+		return nil
+	}
+
+	nginx, err := c.nginxInformer.Lister().NGINXs(namespace).Get(name)
+
+	if err != nil {
+		if errors.IsNotFound(err) {
+			runtime.HandleError(fmt.Errorf("nginx '%s' in work queue no longer exists", key))
+			return nil
+		}
+
+		return err
+	}
+
+	deploymentName := nginx.Spec.DeploymentName
+	if deploymentName == "" {
+		// we don't return an error here so that the item doesn't get requeued
+		// until it is next modified
+		runtime.HandleError(fmt.Errorf("%s: deployment name must be specified", key))
+		return nil
+	}
+
+	deployment, err := c.deploymentsInformer.Lister().Deployments(nginx.Namespace).Get(deploymentName)
+	if errors.IsNotFound(err) {
+		deployment, err = c.clientset.AppsV1beta1().Deployments(nginx.Namespace).Create(newDeployment(nginx))
+	}
+
+	if err != nil {
+		return err
+	}
+
+	return c.updateNGINXStatus(nginx, deployment)
+}
+
+func (c *Controller) updateNGINXStatus(nginx *example.NGINX, deployment *apps.Deployment) error {
+	update := nginx.DeepCopy()
+	update.Status.AvailableReplicas = deployment.Status.AvailableReplicas
+	_, err := c.exampleclientset.ExampleV1alpha1().NGINXs(nginx.Namespace).Update(update)
+	return err
+}
+
+func (c *Controller) enqueueNGINX(obj interface{}) {
+	var key string
+	var err error
+	if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	c.workqueue.Add(key)
+}
+
+func (c *Controller) handleObject(obj interface{}) {
+	var object metav1.Object
+	var ok bool
+	if object, ok = obj.(metav1.Object); !ok {
+		glog.Errorf("error decoding object, invalid type")
+		return
+	}
+	glog.V(4).Infof("Processing object: %s", object.GetName())
+	if ownerRef := metav1.GetControllerOf(object); ownerRef != nil {
+		nginx, err := c.nginxInformer.Lister().NGINXs(object.GetNamespace()).Get(ownerRef.Name)
+
+		if err != nil {
+			glog.V(4).Infof("ignoring orphaned object '%s' of nginx '%s'", object.GetSelfLink(), ownerRef.Name)
+			return
+		}
+
+		c.enqueueNGINX(nginx)
+		return
+	}
+}
+
+func newDeployment(nginx *example.NGINX) *apps.Deployment {
+	return &apps.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      nginx.Spec.DeploymentName,
+			Namespace: nginx.Namespace,
+			OwnerReferences: []metav1.OwnerReference{
+				*metav1.NewControllerRef(nginx, schema.GroupVersionKind{
+					Group:   example.SchemeGroupVersion.Group,
+					Version: example.SchemeGroupVersion.Version,
+					Kind:    "NGINX",
+				}),
+			},
+		},
+		Spec: apps.DeploymentSpec{
+			Replicas: nginx.Spec.Replicas,
+			Template: core.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						"app":        "nginx",
+						"controller": nginx.Name,
+					},
+				},
+				Spec: core.PodSpec{
+					Containers: []core.Container{
+						{
+							Name:  "nginx",
+							Image: "nginx:latest",
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func kubeConfig() (*rest.Config, error) {
```

Comment:
can we use something like this?
https://github.com/kubernetes/apiextensions-apiserver/blob/2bfb07d318ed549813240d1165fcebad6250c666/examples/client-go/main.go#L41

I think that's how it's used in other examples as well.

Reply:
👍  done